provider/core/src/marker.rs

```diff
@@ -108,6 +108,99 @@ impl<M: DataMarker + Sized> DataMarkerExt for M {
+    type EncodeAsVarULE: 'a;
+
+    /// If `self` can be wholy represented as [`Self::VarULE`],
+    /// returns it as a reference.
+    ///
+    /// Otherwise, returns `None`.
+    fn maybe_as_varule(&'a self) -> Option<Self::EncodeAsVarULE>;
+
+    /// Converts from `Self::EncodeAsVarULE` to `Self`.
+    fn from_varule(varule: Self::EncodeAsVarULE) -> Self;
+}
+
+/// Implements required traits on data structs, such as [`MaybeExportAsVarULE`].
+#[macro_export]
+macro_rules! __data_struct {
+    (<$generic:ident: $bound:tt> $ty:path) => {
+        impl<$generic: $bound> $crate::marker::MaybeExportAsVarULE<'_> for $ty {
+            type VarULE = $crate::marker::NeverVarULE;
+            type EncodeAsVarULE = $crate::marker::NeverVarULE;
+            fn maybe_as_varule(&self) -> Option<Self::VarULE> {
+                None
+            }
+            fn from_varule(input: Self::VarULE) -> Self {
+                match input {}
+            }
+        }
+    };
+    ($ty:path) => {
+        impl $crate::marker::MaybeExportAsVarULE<'_> for $ty {
+            type VarULE = $crate::marker::NeverVarULE;
+            type EncodeAsVarULE = $crate::marker::NeverVarULE;
+            fn maybe_as_varule(&self) -> Option<Self::VarULE> {
+                None
+            }
+            fn from_varule(input: Self::VarULE) -> Self {
+                match input {}
+            }
+        }
+    };
+    ($ty:path, varule: $varule:path, as_varule: $as_varule:path, from_varule: $from_varule:path) => {
+        impl<'a> $crate::marker::MaybeExportAsVarULE<'a> for $ty {
+            type VarULE = $varule;
+            type EncodeAsVarULE = &'a $varule;
+            fn maybe_as_varule(&'a self) -> Option<Self::EncodeAsVarULE> {
+                Some($as_varule(self))
+            }
+            fn from_varule(input: Self::EncodeAsVarULE) -> Self {
+                $from_varule(input)
+            }
+        }
+    };
+}
+#[doc(inline)]
+pub use __data_struct as data_struct;
+
+/// An empty enum used for implementations of [`MaybeExportAsVarULE`]
+/// that do not use the VarULE storage optimization.
+#[derive(Debug)]
+#[allow(clippy::exhaustive_enums)] // empty enum
+pub enum NeverVarULE {}
+
+/// Safety checklist for VarULE:
+///
+/// 1. No uninitialized or padding bytes (empty type)
+/// 2. Type is an empty enum
+/// 3. `VarULE::validate_bytes()` always returns errors
+/// 4. `VarULE::validate_bytes()` always returns errors
+/// 5. `VarULE::from_bytes_unchecked()` returns the same pointer
+/// 6. All other methods are left with their default impl
+/// 7. Byte equality is semantic equality
+unsafe impl VarULE for NeverVarULE {
+    fn validate_bytes(_bytes: &[u8]) -> Result<(), UleError> {
+        Err(UleError::parse::<Self>())
+    }
+    unsafe fn from_bytes_unchecked(bytes: &[u8]) -> &Self {
+        // Note: this returns a thin pointer instead of a fat pointer
+        &*(bytes.as_ptr() as *const Self)
+    }
+}
```

Comment:
@Manishearth Flagging this code for a more detailed unsafe review.

Reply:
This shouldn't be an empty enum IMO, there are far too many tricky things there. Let's just use a ZST and have it always error.

Reply:
The empty enum is required in order to implement @robertbastian's preferred solution (discussed in the meeting this morning) where `from_varule` is always implemented; the empty enum means that types that don't support VarULE can match on it. However, I would prefer a different set of traits.

Reply:
Are you concerned about `from_bytes_unchecked` taking in a fat pointer and returning a thin pointer?

Reply:
The unsafe review for zerovec becomes _much_ trickier when you start considering uninhabited types, since it's insta-UB to have an `Empty`, `&Empty`, or other such types floating around. We have not reviewed the zerovec crate for this, and I am unwilling to perform such a deep review and updation of invariants for this change.

Generally uninhabited types do not mix well with unsafe code. It's typical to assume they won't be introduced around this type of unsafe code when reviewing it, and opting uninhabited types into unsafe traits is a major red flag that requires reevaluation of this assumption.


`validate_bytes` always failing makes this _plausibly_ correct, but I'm not able to quickly verify that in a nonlocal way. We have a lot of code using these traits.

Reply:
IMO just abort instead of matching on the empty enum.

Reply:
Yes, I will remove the condition and impl for uninhabited types.

> Are you concerned about `from_bytes_unchecked` taking in a fat pointer and returning a thin pointer?

^ can you address this?

Reply:
To clarify: I can do either of the following

1. Implement `ULE` on a ZST like `()`, and use `[()]` as the placeholder VarULE type (this is what I did in an earlier version of this PR)
2. Implement `VarULE` on the ZST and use `()` directly as the placeholder VarULE, but then it is a thin pointer.

Reply:
Oh, I hadn't noticed that had changed as well. I am somewhat concerned about that, we do assume VarULE types are DSTs but I don't actually think we document it as a requirement on the trait, or have unsafe code relying on it.

So it should be fine. But the `[ZST]` approach is also good.